jgrapht-core/src/main/java/org/jgrapht/traverse/DepthFirstIterator.java

```diff
@@ -34,42 +34,14 @@
  * @param <V> the graph vertex type
  * @param <E> the graph edge type
  *
- * @author Liviu Rau
- * @author Barak Naveh
- * @since Jul 29, 2003
+ * @author Joris Kinable
  */
 public class DepthFirstIterator<V, E>
-    extends CrossComponentIterator<V, E, DepthFirstIterator.VisitColor>
+    extends CrossComponentIterator<V, E, DepthFirstIterator.SearchNodeData>
 {
-    /**
-     * Sentinel object. Unfortunately, we can't use null, because ArrayDeque won't accept those. And
-     * we don't want to rely on the caller to provide a sentinel object for us. So we have to play
-     * typecasting games.
-     */
-    public static final Object SENTINEL = new Object();
-
-    /**
-     * Standard vertex visit state enumeration.
-     */
-    protected static enum VisitColor
-    {
-        /**
-         * Vertex has not been returned via iterator yet.
-         */
-        WHITE,
-
-        /**
-         * Vertex has been returned via iterator, but we're not done with all of its out-edges yet.
-         */
-        GRAY,
-
-        /**
-         * Vertex has been returned via iterator, and we're done with all of its out-edges.
-         */
-        BLACK
-    }
 
-    private Deque<Object> stack = new ArrayDeque<>();
+    //private Stack<V> stack = new Stack<>();
+    private Stack<StackEntry> stack = new Stack<>();
```

Comment:
Probably `ArrayDeque` I think.

`LinkedList` uses more memory per element than array-based data structures if I understand correctly (although, thinking more about it, this may not actually be true, since array-based data structures by their nature need to grow their internal arrays if they get too full, meaning array-based data structures are usually larger than necessary at any given point in time).

Furthermore I would guess that `LinkedList` is slower on modern hardware, since it uses pointers a lot, which I believe is not CPU-cache-friendly.

So I'm inclined to go for `ArrayDeque` unless profiling shows that `LinkedList` is better.